//----------------------------------------------------------------------------------------------------------------------
/// @file MainWindow.h
/// @brief Dialog window that handles user interaction and user interface design to interact with the application
/// @author Michelle Wu
/// @version 1.0
//----------------------------------------------------------------------------------------------------------------------

#ifndef MAINWINDOW_H
#define MAINWINDOW_H

#include <QFileDialog>
#include <QMainWindow>

#include "BlendDialog.h"
#include "NGLScene.h"


namespace Ui
{
  class MainWindow;
}

class MainWindow : public QMainWindow
{
  Q_OBJECT

public:

  //-------------------------------------------------------------------------------
  /// @brief Ctor
  //-------------------------------------------------------------------------------
  explicit MainWindow( QWidget* _parent = 0 );
  //-------------------------------------------------------------------------------
  /// @brief Dtor
  //-------------------------------------------------------------------------------
  ~MainWindow();
  //-------------------------------------------------------------------------------

private slots:

  //-------------------------------------------------------------------------------
  /// @brief Slot to indicate the load file button has been clicked
  //-------------------------------------------------------------------------------
  void loadButtonClicked();
  //-------------------------------------------------------------------------------
  /// @brief Play animation
  //-------------------------------------------------------------------------------
  void playButtonClicked();
  //-------------------------------------------------------------------------------
  /// @brief Pause animation
  //-------------------------------------------------------------------------------
  void pauseButtonClicked();
  //-------------------------------------------------------------------------------
  /// @brief Play animation backward
  //-------------------------------------------------------------------------------
  void backwardButtonClicked();
  //-------------------------------------------------------------------------------
  /// @brief Play animation forward
  //-------------------------------------------------------------------------------
  void forwardButtonClicked();
  //-------------------------------------------------------------------------------
  /// @brief When selecting items in drop down menu
  /// @param [in] _changedIndex Changed index
  //-------------------------------------------------------------------------------
  void indexChanged( const int &_changedIndex );
  //-------------------------------------------------------------------------------
  /// @brief Play next frame
  //-------------------------------------------------------------------------------
  void nextFrameButtonClicked();
  //-------------------------------------------------------------------------------
  /// @brief Play previous frame
  //-------------------------------------------------------------------------------
  void previousFrameButtonClicked();
  //-------------------------------------------------------------------------------
  /// @brief Perform interpolation between frames
  //-------------------------------------------------------------------------------
  void blendButtonClicked();
  //-------------------------------------------------------------------------------
  /// @brief Reset animation from frame 0
  //-------------------------------------------------------------------------------
  void resetButtonClicked();
  //-------------------------------------------------------------------------------

private:

  //-------------------------------------------------------------------------------
  /// @brief User interface
  //-------------------------------------------------------------------------------
  Ui::MainWindow* m_ui;
  //-------------------------------------------------------------------------------
  /// @brief OpenGL widget (our scene)
  //-------------------------------------------------------------------------------
  NGLScene* m_glWidget;
  //-------------------------------------------------------------------------------
  /// @brief Blend/interpolation dialog window
  //-------------------------------------------------------------------------------
  BlendDialog* m_blendDialog;
  //-------------------------------------------------------------------------------

};

#endif // MAINWINDOW_H
